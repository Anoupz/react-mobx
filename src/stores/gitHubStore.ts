import asyncComputed from '../utils/asyncComputed';

export class GitHubStore {
  trendingJavascriptRepos = asyncComputed(async () => {
    const response = await fetch(`https://github-trending-api.now.sh/repositories?language=javascript&since=weekly`);
    const result = await response.json();
    return result;
  });
}

export default new GitHubStore();
