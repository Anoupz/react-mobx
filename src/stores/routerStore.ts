import { computed, reaction } from 'mobx';
import createBrowserHistory from 'mobx-history/createBrowserHistory';

const UrlPattern = require('url-pattern');
const _mainRoute = '/main';
export const mainPageRoute = new UrlPattern(_mainRoute + '(/*)');

const _gitRoute = '/trendingGitRepo';
export const gitPageRoute = new UrlPattern(_gitRoute + '(/*)');

const routes = [mainPageRoute, gitPageRoute];

export type PAGE = 'main' | 'trendingGitRepo';
export class RouterStore {
  history = createBrowserHistory();

  redirectReaction = reaction(
    () => this.activeRoutes.length,
    () => {
      if (this.activeRoutes.length === 0) {
        const searchParam = new URL(window.location.href).search;
        this.history.push({
          pathname: mainPageRoute.stringify({ page: '/' }),
          search: searchParam
        });
      }
    },
    { fireImmediately: true }
  );

  @computed
  get activeRoutes() {
    return routes.filter(route => !!route.match(this.history.location.pathname));
  }
}
export default new RouterStore();
