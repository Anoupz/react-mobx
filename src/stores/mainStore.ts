import { observable, action } from 'mobx';
import { RouterStore } from './routerStore';

export class MainStore {
  @observable headerText: string = 'React';

  constructor(private _routerStore: RouterStore) {}

  @action
  updateHeaderTxt(newString: string) {
    this.headerText = newString;
  }

  @action
  goToTrendingGithubPage() {
    this._routerStore.history.push(`/trendingGitRepo`);
  }
}
