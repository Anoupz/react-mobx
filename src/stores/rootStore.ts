import { inject } from 'mobx-react';

import { TypedInject } from './../utils/typedInject';
import { MainStore } from './mainStore';
import routerStore from './routerStore';
import gitHubStore from './gitHubStore';

export const mainStore = new MainStore(routerStore);
export const rootStores = {
  routerStore,
  mainStore,
  gitHubStore
};

export type Stores = typeof rootStores;

export const typedInject = inject as TypedInject<Stores>;
