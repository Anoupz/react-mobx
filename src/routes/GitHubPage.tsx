import React, { Component } from 'react';
import { Dimmer, Loader, Message, Grid, Header, Table } from 'semantic-ui-react';
import { observer } from 'mobx-react';

import { typedInject } from '../stores/rootStore';
import { GitHubStore } from '../stores/gitHubStore';

interface Props {
  gitHubStore: GitHubStore;
}

@observer
export class GitHubPage extends Component<Props, {}> {
  render() {
    const { trendingJavascriptRepos } = this.props.gitHubStore;
    return (
      <Grid width={8} style={{ marginTop: '6rem', marginBottom: '6rem' }}>
        <Header>Popular Javascript repo</Header>
        {trendingJavascriptRepos.case({
          pending: () => (
            <Dimmer active inverted>
              <Loader size="mini">Loading details...</Loader>
            </Dimmer>
          ),
          rejected: () => (
            <Message negative>
              <Message.Header>We encountered some errors.</Message.Header>
              <p>We had a problem loading this page. Please try again later.</p>
            </Message>
          ),
          fulfilled: data => (
            <Table color="purple">
              <Table.Header>
                <Table.Row textAlign="center">
                  <Table.HeaderCell>Name</Table.HeaderCell>
                  <Table.HeaderCell>Repo Url</Table.HeaderCell>
                  <Table.HeaderCell>Stars</Table.HeaderCell>
                </Table.Row>
              </Table.Header>
              <Table.Body>
                {data.map((repo, index) => (
                  <Table.Row key={index} textAlign="center">
                    <Table.Cell width={6}>{repo.name}</Table.Cell>
                    <Table.Cell width={6}>{repo.url}</Table.Cell>
                    <Table.Cell width={4}>{repo.stars}</Table.Cell>
                  </Table.Row>
                ))}
              </Table.Body>
            </Table>
          )
        })}
      </Grid>
    );
  }
}

export default typedInject('gitHubStore')(GitHubPage);
