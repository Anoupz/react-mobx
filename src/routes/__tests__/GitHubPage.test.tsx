import * as React from 'react';
import { shallow } from 'enzyme';

import { GitHubStore } from '../../stores/gitHubStore';
import GitHubPage from '../GitHubPage';

describe('Git Hub Page testing', () => {
  let gitHubStore: GitHubStore = new GitHubStore();

  it('renders without crashing', () => {
    shallow(<GitHubPage gitHubStore={gitHubStore} />);
  });

  it('renders correctly', () => {
    expect(shallow(<GitHubPage gitHubStore={gitHubStore} />)).toMatchSnapshot();
  });
});
