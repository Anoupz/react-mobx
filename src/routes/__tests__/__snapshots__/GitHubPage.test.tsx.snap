// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Git Hub Page testing renders correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <inject-GitHubPage-with-gitHubStore
    gitHubStore={
      GitHubStore {
        "trendingJavascriptRepos": AsyncComputed {
          "computeFn": [Function],
          "refreshCallCount": 0,
          Symbol(mobx did run lazy initializers): true,
          Symbol(mobx administration): ObservableObjectAdministration$$1 {
            "defaultEnhancer": [Function],
            "keysAtom": Atom$$1 {
              "diffValue": 0,
              "isBeingObserved": false,
              "isPendingUnobservation": false,
              "lastAccessedBy": 0,
              "lowestObserverState": 2,
              "name": "AsyncComputed@15.keys",
              "observers": Set {},
            },
            "name": "AsyncComputed@15",
            "target": [Circular],
            "values": Map {
              "pending" => true,
              "rejected" => false,
              "fulfilled" => false,
              "refreshCallCount" => 0,
              "state" => "pending",
              "value" => undefined,
              "error" => undefined,
              "_internalObservable" => Promise {
                "case": [Function],
                "isPromiseBasedObservable": true,
                "state": "pending",
                "value": undefined,
                Symbol(mobx administration): ObservableObjectAdministration$$1 {
                  "defaultEnhancer": [Function],
                  "keysAtom": Atom$$1 {
                    "diffValue": 0,
                    "isBeingObserved": false,
                    "isPendingUnobservation": false,
                    "lastAccessedBy": 0,
                    "lowestObserverState": 2,
                    "name": "Promise@29.keys",
                    "observers": Set {},
                  },
                  "name": "Promise@29",
                  "target": [Circular],
                  "values": Map {
                    "value" => undefined,
                    "state" => "pending",
                  },
                },
              },
            },
          },
        },
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "gitHubStore": GitHubStore {
        "trendingJavascriptRepos": AsyncComputed {
          "computeFn": [Function],
          "refreshCallCount": 0,
          Symbol(mobx did run lazy initializers): true,
          Symbol(mobx administration): ObservableObjectAdministration$$1 {
            "defaultEnhancer": [Function],
            "keysAtom": Atom$$1 {
              "diffValue": 0,
              "isBeingObserved": false,
              "isPendingUnobservation": false,
              "lastAccessedBy": 0,
              "lowestObserverState": 2,
              "name": "AsyncComputed@15.keys",
              "observers": Set {},
            },
            "name": "AsyncComputed@15",
            "target": [Circular],
            "values": Map {
              "pending" => true,
              "rejected" => false,
              "fulfilled" => false,
              "refreshCallCount" => 0,
              "state" => "pending",
              "value" => undefined,
              "error" => undefined,
              "_internalObservable" => Promise {
                "case": [Function],
                "isPromiseBasedObservable": true,
                "state": "pending",
                "value": undefined,
                Symbol(mobx administration): ObservableObjectAdministration$$1 {
                  "defaultEnhancer": [Function],
                  "keysAtom": Atom$$1 {
                    "diffValue": 0,
                    "isBeingObserved": false,
                    "isPendingUnobservation": false,
                    "lastAccessedBy": 0,
                    "lowestObserverState": 2,
                    "name": "Promise@36.keys",
                    "observers": Set {},
                  },
                  "name": "Promise@36",
                  "target": [Circular],
                  "values": Map {
                    "value" => undefined,
                    "state" => "pending",
                  },
                },
              },
            },
          },
        },
      },
    },
    "ref": [Function],
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "gitHubStore": GitHubStore {
          "trendingJavascriptRepos": AsyncComputed {
            "computeFn": [Function],
            "refreshCallCount": 0,
            Symbol(mobx did run lazy initializers): true,
            Symbol(mobx administration): ObservableObjectAdministration$$1 {
              "defaultEnhancer": [Function],
              "keysAtom": Atom$$1 {
                "diffValue": 0,
                "isBeingObserved": false,
                "isPendingUnobservation": false,
                "lastAccessedBy": 0,
                "lowestObserverState": 2,
                "name": "AsyncComputed@15.keys",
                "observers": Set {},
              },
              "name": "AsyncComputed@15",
              "target": [Circular],
              "values": Map {
                "pending" => true,
                "rejected" => false,
                "fulfilled" => false,
                "refreshCallCount" => 0,
                "state" => "pending",
                "value" => undefined,
                "error" => undefined,
                "_internalObservable" => Promise {
                  "case": [Function],
                  "isPromiseBasedObservable": true,
                  "state": "pending",
                  "value": undefined,
                  Symbol(mobx administration): ObservableObjectAdministration$$1 {
                    "defaultEnhancer": [Function],
                    "keysAtom": Atom$$1 {
                      "diffValue": 0,
                      "isBeingObserved": false,
                      "isPendingUnobservation": false,
                      "lastAccessedBy": 0,
                      "lowestObserverState": 2,
                      "name": "Promise@43.keys",
                      "observers": Set {},
                    },
                    "name": "Promise@43",
                    "target": [Circular],
                    "values": Map {
                      "value" => undefined,
                      "state" => "pending",
                    },
                  },
                },
              },
            },
          },
        },
      },
      "ref": [Function],
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
